asllib/AST.mli

```diff
@@ -201,6 +203,9 @@ type stmt_desc =
   | S_Cond of expr * stmt * stmt
   | S_Case of expr * case_alt list
   | S_Assert of expr
+  | S_For of identifier * expr * for_direction * expr * stmt
+  | S_While of expr * stmt
+  | S_Repeat of stmt * expr
```

Comment:
or should `e` be in `s`'s scope?

Reply:
> Should we keep `S_Repeat (s, e)` as a sugar for `S_Then (s, S_While (e, s))`? I don't really mind but it makes the interpreter/typer more complex for nothing?

I am not an adept of syntactic sugar, as it means departing from the original source, which often produces cryptic error messages, among other difficulties (here we would also duplicate code). Moreover, code for implementing `repeat' remains simple.

Reply:
> or should `e` be in `s`'s scope? I'm not sure that would be something that we want to

I am not sure I understand. What scope? I was not aware that the while or repeat loops define scope.

Reply:
> > Should we keep `S_Repeat (s, e)` as a sugar for `S_Then (s, S_While (e, s))`? I don't really mind but it makes the interpreter/typer more complex for nothing?
> 
> I am not an adept of syntactic sugar, as it means departing from the original source, which often produces cryptic error messages, among other difficulties (here we would also duplicate code). Moreover, code for implementing `repeat' remains simple.

I don't understand why it would duplicate code? As the pattern for `S_Repeat` could be `assert false` (or some other error)? What I suggest is at compile-time, for example in `annotate_stmt`, we would simply rewrite `S_Repeat` into `S_While`.

The main problem I have is that this interpreter is meant to be read, and it getting bigger means that 1) we need to explain more things to people, 2) people will get scared by its size, and 3) we have more lines to maintain.

For the error messages, we don't have any error message that deals with while-specific code, but in that case we could rephrase it with "loop" instead of "while".

Reply:
> > or should `e` be in `s`'s scope? I'm not sure that would be something that we want to
> 
> I am not sure I understand. What scope? I was not aware that the while or repeat loops define scope.

If I understand correctly, the body of the loop is an "enclosing block" of any statement inside its body. For example the following is illegal:
```
var i = 0;
while i < 10;
   let j = 10 - i;
   i = i + j / 2 + 1;
end
let x = j; // Illegal
```

The precise question is the legality of:

```
repeat
   Mem[address] = 1;
   let data = Mem[address];
until data == 1;
```

Reply:
> Ok I see now, The code for implementing loop unrolling limitation (a.k.a. "Too Far") introduces some complexity and is now shared in the interpreter, see function [`eval_loop`](https://github.com/herd/herdtools7/blob/a1aba97628dc3824cef7409b1da5cb019572c92b/asllib/Interpreter.ml#L703).

Great! That does solve my issues for while/repeat.

> As regards scopes, if we follow C, the scope of a repeat loop body does not include the loop condition. In that aspect the while and repeat loops would not differ.
>
> Moreover scopes are not handled for the moment. In particular, variable declaration and asigements are not distinguished. Do you have plans?

Yes! See for example https://github.com/herd/herdtools7/pull/595/commits/c38c92c616992a061408161a8646942467ebf41b

> Of course, one can replace `repeat` loops by `while` loops earlier (BTW the condition is negated). I am not sure this a better solution.

Given the small but quite significant difference when loop unrolling, I think your current changes are probably the best. Let's keep those on that front.